### PR TITLE
improvements to the keycloak_realm_key module

### PR DESCRIPTION
**Pull request by gbolo**
##### SUMMARY
This PR makes the following improvements to the `keycloak_realm_key` module:

- adds support for RSA enc key usage `rsa-enc`
- adds support for an additional 8 key algorithms
- fixes a bug where the user must specify a certificate (no longer required)
- removes `no_log` for certificate since this is not sensitive information

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
module: `keycloak_realm_key`

##### ADDITIONAL INFORMATION
Tested against Keycloak `v22.0.4` API.

### CODE REVIEW

File: changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+minor_changes:
+  - keycloak_realm_key module now supports rsa enc key usage (`rsa-enc`) via `provider_id` parameter (https://github.com/ansible-collections/community.general/pull/7698).
```
Comment (felixfontein) [inline changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml:2]: Please note that the collection uses the format `<module_name> - <desc> (URL).`. Also note that formatting inside the lines should be RST, not MarkDown, so you need double backticks for code formatting.
```suggestion
  - keycloak_realm_key - the ``provider_id`` option now supports RSA encryption key usage (value ``rsa-enc``) (https://github.com/ansible-collections/community.general/pull/7698).
```
Similar for the other lines.

File: plugins/modules/keycloak_realm_key.py [hunk 2]
```diff
@@ -103,7 +103,7 @@
                 description:
                     - Key algorithm.
```
Comment (felixfontein) [inline plugins/modules/keycloak_realm_key.py:104]:
```suggestion
                    - Key algorithm.
                    - The values V(RS384), V(RS512), ..., V(RSA-OAEP-256) have been added in community.general 8.2.0.
```
(add the remaining ones in there)

File: plugins/modules/keycloak_realm_key.py [hunk 3]
```diff
@@ -74,7 +74,7 @@
     provider_id:
         description:
             - The name of the "provider ID" for the key.
```
Comment (felixfontein) [inline plugins/modules/keycloak_realm_key.py:76]: Similar here.

File: changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+  - keycloak_realm_key module now supports 8 additional key algorithms via `config.algorithm` parameter (https://github.com/ansible-collections/community.general/pull/7698).
+  - keycloak_realm_key module now sets a default empty string for `config.certificate` parameter instead of forcing user to specify a value (https://github.com/ansible-collections/community.general/pull/7698).
```
Comment (felixfontein) [inline changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml:4]: What's the rationale behind this change? In the PR description you mention:

> fixes a bug where the user must specify a certificate (no longer required)

but I don't see how this is a bug?

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

File: changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml (hunk 4 above)
Comment (gbolo) [reply changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml:4]: Hi @felixfontein, the reason that I considered it a bug is because the very first example listed suggests that it is not required (so perhaps an error in docs):
```
EXAMPLES = '''
- name: Manage Keycloak realm key (certificate autogenerated by Keycloak)
  community.general.keycloak_realm_key:
    name: custom
    state: present
    parent_id: master
    provider_id: rsa
    auth_keycloak_url: http://localhost:8080/auth
    auth_username: keycloak
    auth_password: keycloak
    auth_realm: master
    config:
      private_key: "{{ private_key }}"
      enabled: true
      active: true
      priority: 120
      algorithm: RS256
```

**Comment by danekja** [comment]
If I may (not a maintainer, merely a user of the collection) - I would propose to cover the whole set of supported algorithms by tests. And thanks for contribution, we were going to implement this addition ourselves in the following months as well.

File: changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml (hunk 4 above)
Comment (felixfontein) [reply changelogs/fragments/7698-improvements-to-keycloak_realm_key.yml]: The problem is that adding a default makes it pretty hard to change the behavior of not specifying the value later - for example so that not specifying a value doesn't change the certificate if the key doesn't change.

It's better to fix the docs instead (best done in a separate PR so it can be backported further - features can only end up in the latest `stable-x` branch, but docs fixes can go back to all maintained `stable-x` branches.).

**Comment by felixfontein** [comment]
Some feedback from the module maintainers would be nice.

In general the change looks OK to me, except that `default` / removal of `required=true`. The module simply passes the values on to the API, so I don't think unit tests are needed. (Integration tests would be nice, but these don't run in CI, so their value is limited.)

File: plugins/modules/keycloak_realm_key.py [hunk 5]
```diff
@@ -120,7 +123,7 @@
                       key must match O(config.algorithm) and O(provider_id).
                     - If you want Keycloak to automatically generate a certificate using your private key
                       then set this to an empty string.
-                required: true
+                default: ''
```
Comment (mattock) [inline plugins/modules/keycloak_realm_key.py:126]: I can't remember the details of why this is/was set to _required: true_. Maybe the idea was to be explicit about "I don't want to pass a certificate of my own" instead of implicitly assuming it. If the integration tests do not fail then I'm not opposed to this change, though.

File: plugins/modules/keycloak_realm_key.py [hunk 6]
```diff
@@ -244,16 +247,30 @@ def main():
...
+                    choices=[
+                        "RS256",
+                        "RS384",
+                        "RS512",
+                        "PS256",
+                        "PS384",
+                        "PS512",
+                        "RSA1_5",
+                        "RSA-OAEP",
+                        "RSA-OAEP-256",
```
Comment (mattock) [inline plugins/modules/keycloak_realm_key.py:269]: Adding these makes sense. I only had a need for RS256 and I did not want to overreach. However, I think adding one integration test per algorithm would be useful as the list of supported algorithms might change over time. Or these might be a typo in there :smile:.

Comment (gbolo) [reply plugins/modules/keycloak_realm_key.py:269]: Yea I can add some basic integration tasks to check these I guess, although some algorithms belong on certain key types only. For example, you would not want to attempt to set `RSA-OAEP` on your signing key. I was hoping that the end user of this module would understand the values they are setting so that we don't have to write too much validation code. Ideally the keycloak api will return an error in these cases so we can just pass that error on. I think maybe we should tackle this in another PR though...

File: plugins/modules/keycloak_realm_key.py (hunk 5 above)
Comment (gbolo) [reply plugins/modules/keycloak_realm_key.py:126]: I checked out this API call today and I can confirm that if the cert is an empty string then keycloak will generate a new one. If the cert is omitted entirely from the json object then keycloak will leave the cert alone. @felixfontein it would seem like we should address this in another PR because neither this PR or the existing functionality seems correct here. I can remove the default to match the existing behaviour. I would imagine that most people would probably like to manage their own cert along with the key though...

Comment (felixfontein) [reply plugins/modules/keycloak_realm_key.py]: There is no "correct", there is only no "all possible use-cases are supported". I would either try to support all possible use-cases (which includes "keep the certificate as-is"), or not change it (which the PR in its latest form does).

I agree, supporting the "keep the certificate as-is" use-case should better be done in a separate PR.

Comment (gbolo) [reply plugins/modules/keycloak_realm_key.py]: Excellent assessment. Let's get this merged and move on then.

**Review by felixfontein — APPROVED** [review]
Looks good to me. If nobody objects, I'll merge this in ~a week.

**Comment by felixfontein** [comment]
@gbolo thanks for your contribution!
@danekja @mattock thanks for reviewing!
